docs/grpc/third_party_stubs_design.md

```diff
@@ -0,0 +1,305 @@
+│   │   └── ...
+│   └── ...
+├── common/
+│   └── grpc_stubs/                        # Generated stubs for all protos
+│       ├── openconfig/
+│       │   ├── gnmi/
+│       │   └── gnoi/
+├── build_grpc_stubs.sh                    # Unified build script
+└── ...
+```
+
+### Git Submodules
+
+Git submodules will be used to maintain specific versions of third-party proto definitions. This approach provides:
+
+1. **Version Control**: Each submodule can be pinned to a specific commit/tag
+2. **Reproducibility**: Tests will always use the same version of dependencies
+3. **Offline Operation**: No network connectivity required during test execution
+4. **Update Process**: Clear mechanism for updating dependencies when needed
+
+Example of adding a submodule:
+```bash
+git submodule add https://github.com/openconfig/gnmi.git tests/third_party/openconfig/gnmi
+git submodule add https://github.com/openconfig/gnoi.git tests/third_party/openconfig/gnoi
+```
+
+### Unified Build Script
+
+A single build script (`build_grpc_stubs.sh`) will generate stubs for all proto definitions. The script will:
+
+1. Take a base directory as input (typically the test directory)
+2. Generate stubs for all proto files in all submodules
+3. Organize the output in a standardized directory structure
+4. Create necessary `__init__.py` files for Python package imports
+
+```bash
+#!/bin/bash
+
+# This script generates gRPC stubs for all third-party proto definitions
+
+set -e
+set -u
+set -o pipefail
+
+if [ "$#" -ne 1 ]; then
+    echo "Usage: $0 <base_directory>"
+    exit 1
+fi
+
+BASE_DIR="$1"
+THIRD_PARTY_DIR="$BASE_DIR/third_party"
+GENERATED_DIR="$BASE_DIR/common/grpc_stubs"
+
+# Clean generated directory
+if [ -d "$GENERATED_DIR" ]; then
+    echo "Removing existing directory: $GENERATED_DIR"
+    rm -rf "$GENERATED_DIR"
+fi
+echo "Creating directory: $GENERATED_DIR"
+mkdir -p "$GENERATED_DIR"
+
+# Function to generate stubs for a proto file
+generate_stubs() {
+    local proto_file="$1"
+    local proto_dir=$(dirname "$proto_file")
+    local proto_path="$THIRD_PARTY_DIR"
+
+    echo "Generating stubs for $proto_file"
+    python -m grpc_tools.protoc \
+        --proto_path="$proto_path" \
+        --python_out="$GENERATED_DIR" \
+        --grpc_python_out="$GENERATED_DIR" \
+        "$proto_file"
+}
+
+# Find all proto files
+find "$THIRD_PARTY_DIR" -name "*.proto" | while read proto_file; do
+    generate_stubs "$proto_file"
+done
+
+# Create __init__.py files for all directories
+find "$GENERATED_DIR" -type d | while read dir; do
+    touch "$dir/__init__.py"
+done
+
+# Handle path corrections for grpc compiler bug if needed
+# (similar to the existing build-gnmi-stubs.sh script)
+
+echo "gRPC stubs generated successfully."
+```
+
+### Integration with pytest
+
+The build script will be invoked during pytest session setup through a session-scoped fixture in `conftest.py`:
+
+```python
+@pytest.fixture(scope="session", autouse=True)
```

Comment:
I think if it's autouse, then fixtures that don't need this fixture will still have this fixture be run?